File: 139-word-break/word-break.go
func canBreak(wordSet map[string]bool, index int, s *string, memo map[int]bool) bool{
    // memoized check
    if val, ok := memo[index]; ok {
        return val
    }
    // found word
    if index == len(*s) {
        return true
    }

    // check current word
    for i := index; i < len(*s); i++{
        if !wordSet[(*s)[index: i + 1]]{
            continue
        }

        if canBreak(wordSet, i + 1, s, memo){
            memo[index] = true
            return true
        }
    }

    memo[index] = false
    return false
}

func wordBreak(s string, wordDict []string) bool {
    wordSet := make(map[string]bool)
    for _, word := range wordDict {
        wordSet[word] = true
    }

    memo := make(map[int]bool, len(s) + 1)
    return canBreak(wordSet, 0, &s, memo)

}

File: 139-word-break/word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        word_set = set(wordDict)
        memo = {}

        def can_break(start: int) -> bool:
            if start in memo:
                return memo[start]

            if start >= len(s):
                return True
            
            for i in range(start, len(s)):
                if s[start: i + 1] in word_set:
                    if can_break(i + 1):
                        memo[start] = True
                        return True
            
            memo[start] = False
            return False
        
        return can_break(0)